include/jemalloc/internal/cache_bin.h

```diff
@@ -194,8 +194,15 @@ cache_bin_assert_earlier(cache_bin_t *bin, uint16_t earlier, uint16_t later) {
  * be associated with the position earlier in memory.
  */
 static inline uint16_t
-cache_bin_diff(cache_bin_t *bin, uint16_t earlier, uint16_t later) {
-	cache_bin_assert_earlier(bin, earlier, later);
+cache_bin_diff(cache_bin_t *bin, uint16_t earlier, uint16_t later, bool racy) {
```

Comment:
It looks to me that, we can't have the bool racy here? Isn't it already fixed, w/o the bool racy here? 

The other aspect is, if the assertion fails, it means we have an underflow bug, i.e. the diff calculated isn't correct. IIUC, the bug was that `cache_bin_low_bound_get` was calling `cache_bin_empty_position_get`, while `cache_bin_empty_position_get` itself isn't safe to call remotely (since the stack_head might load twice which results in a wrong empty position), does that match what you saw @Lapenkov ?

Reply:
This is indeed unrelated to the issue reported, but it's for `cache_bin_ncached_get_internal(bin, racy = true)`.

There, we calculate the difference between `(uint16_t)(uintptr_t)stack_head` and `low_bits_empty`. It can happen that at the moment we calculate the difference, `(uint16_t)(uintptr_t)stack_head > low_bits_empty`, but by the time we reach `assert(bin->low_bits_full > bin->low_bits_empty);`, we could've stashed some items and changed `low_bits_full` so that the assertion fails too.

Reply:
Thanks for explaining. I see the point now.

Maybe we should update the assertion in `cache_bin_assert_earlier`? There the `bin->low_bits_full` really was referring to `cache_bin_low_bits_low_bound`. Am I understanding correctly that, if we use the low_bound, we wouldn't have to bypass it for the racy case? (not asking to change to that way, but for sanity checking).

Reply:
Actually, yes, you're right. Instead of skipping the assertion, I can calculate the low bits low bound with read-only variables and use it for comparison. This will make the code easier. Will update the PR tomorrow.

Reply:
Discussed offline. Let's still keep the current way, i.e. skipping the assertion in the remote reading case.